xarray/core/parallel.py

```diff
@@ -0,0 +1,141 @@
+    #    raise ValueError("map_blocks can only be used with DataArrays at present.")
+
+    if isinstance(obj, DataArray):
+        dataset = obj._to_temp_dataset()
+        to_array = True
+    else:
+        dataset = obj
+        to_array = False
+
+    dataset_dims = list(dataset.dims)
+
+    graph = {}
+    gname = "map-%s-%s" % (dask.utils.funcname(func), dask.base.tokenize(dataset))
+
+    # map dims to list of chunk indexes
+    # If two different variables have different chunking along the same dim
+    # .chunks will raise an error.
+    chunks = dataset.chunks
+    ichunk = {dim: range(len(chunks[dim])) for dim in chunks}
+    # mapping from chunk index to slice bounds
+    chunk_index_bounds = {dim: np.cumsum((0,) + chunks[dim]) for dim in chunks}
+
+    # iterate over all possible chunk combinations
+    for v in itertools.product(*ichunk.values()):
+        chunk_index_dict = dict(zip(dataset_dims, v))
+
+        # this will become [[name1, variable1],
+        #                   [name2, variable2],
+        #                   ...]
+        # which is passed to dict and then to Dataset
+        data_vars = []
+        coords = []
+
+        for name, variable in dataset.variables.items():
+            # make a task that creates tuple of (dims, chunk)
+            if dask.is_dask_collection(variable.data):
+                var_dask_keys = variable.__dask_keys__()
+
+                # recursively index into dask_keys nested list to get chunk
+                chunk = var_dask_keys
+                for dim in variable.dims:
+                    chunk = chunk[chunk_index_dict[dim]]
+
+                task_name = ("tuple-" + dask.base.tokenize(chunk),) + v
+                graph[task_name] = (tuple, [variable.dims, chunk])
+            else:
+                # numpy array with possibly chunked dimensions
+                # index into variable appropriately
+                subsetter = dict()
+                for dim in variable.dims:
+                    if dim in chunk_index_dict:
+                        which_chunk = chunk_index_dict[dim]
+                        subsetter[dim] = slice(
+                            chunk_index_bounds[dim][which_chunk],
+                            chunk_index_bounds[dim][which_chunk + 1],
+                        )
+
+                subset = variable.isel(subsetter)
+                task_name = (name + dask.base.tokenize(subset),) + v
+                graph[task_name] = (tuple, [subset.dims, subset])
+
+            # this task creates dict mapping variable name to above tuple
+            if name in dataset.data_vars:
+                data_vars.append([name, task_name])
+            if name in dataset.coords:
+                coords.append([name, task_name])
+
+        graph[(gname,) + v] = (
+            _wrapper,
+            func,
+            (Dataset, (dict, data_vars), (dict, coords), dataset.attrs),
+            to_array,
+            args,
+            kwargs,
+        )
```

Comment:
To make this work on a `Dataset`, we need one more layer to map a dask graph of (list of array) tasks into a list of dask arrays.

Suppose you change `_wrapper` to return a list of arrays Numpy arrays, as I suggested above. Then this would look something like:
```
for i in range(num_data_vars):
    gname_i = gname + '-out' + str(i)
    graph[(gname_i,) + v] = (operator.getitem, (gname,) + v, i)
```
Now you should simply construct the desired dask arrays from each of `gname_i` instead of `gname`.

Reply:
Yes, I got that to work. Though I use  dict to keep track of variable names.